### PR TITLE
Allow checkouts of repositories into non-empty directories

### PR DESCRIPTION
**Pull request by j-walther**
Allow `gix::clone::PrepareFetch` callers to clone and check out into a destination directory that is not empty.

Before this change, `PrepareFetch::new()` unconditionally overrode `create::Options::destination_must_be_empty` to `true` regardless of what the caller passed, so the flag was effectively non-functional and there was no way to opt in to a non-empty destination through the public API.

`gix::create::Options::destination_must_be_empty` now controls this for non-bare repositories:

- `None` (default for `gix::init`) — the directory may already contain files as long as no `.git` directory is present.
- `Some(true)` (effective default for `prepare_clone` / `prepare_clone_bare`) — require an empty destination, preserving the prior clone behavior.
- `Some(false)` — explicitly allow a non-empty destination for clones.

Bare repositories always require an empty destination.

If a clone fails before completion and the prepare handle is dropped, only the `.git` directory created by the clone is removed; any pre-existing user files in the destination are preserved.

### CODE REVIEW

**Review by Byron — CHANGES_REQUESTED** [review]
Thanks, but I think this needs test coverage.
I would also be afraid that it now allows writing non-empty directories, wiping data in the process.

**Comment by j-walther** [comment]
> I would also be afraid that it now allows writing non-empty directories, wiping data in the process.
The current state is that it's effectively impossible to check-out into a non-empty directory making that user-facing flag effectively superfluous.

What would you expect tests for this to look like? Checking out into a non empty directory with the flag both set and not set? The latter would just be a regular checkout.

**Comment by Byron** [comment]
There needs to be a test that fails if the change of this PR isn't present.
Also, there must be a test that shows non-empty directories cannot be checked out into to address

> I would also be afraid that it now allows writing non-empty directories, wiping data in the process.

It might already exist as well, in which case you can point it out to me.

**Comment by j-walther** [comment]
I've added the tests. What I think should be done is to change the `Default` implementation to set `destination_must_be_empty` to `true` by default.

**Review by Byron — CHANGES_REQUESTED** [review]
I don't think I understand the premise. The reason that it won't allow cloning into a non-empty directory is that it will brutally overwrite everything. Git also rejects this, AFAIK.

```
❯ git clone https://github.com/GitoxideLabs/gitoxide ./src/
fatal: destination path './src' already exists and is not an empty directory.
```

The PR title is "Allow checkouts of empty repositories". Thus a test should show that empty repositories couldn't be checked out before, and after the fix they can.

> I've added the tests. What I think should be done is to change the Default implementation to set destination_must_be_empty to true by default.

I agree, and wonder why I didn't that before.

One problem remains: I don't understand what the PR is trying to fix.

**Comment by j-walther** [comment]
> One problem remains: I don't understand what the PR is trying to fix.

Before my changes every code path set the `destination_must_be_empty` flag to `true` in a part of the code the user couldn't influence so you could **never** check out into a non-empty directory. This made the flag essentially meaningless since it'd effectively always be `true`. This PR removes the line that forces it to always be `true` so the user can actually opt into checking out into non-empty directories.

**Comment by j-walther** [comment]
I've updated the tests to better reflect the change and made it so `destination_must_be_empty` is `true` by default.

File: gix/src/clone/mod.rs [hunk 1]
```diff
@@ -99,10 +99,9 @@ impl PrepareFetch {
         mut url: gix_url::Url,
         path: &std::path::Path,
         kind: crate::create::Kind,
-        mut create_opts: crate::create::Options,
+        create_opts: crate::create::Options,
         open_opts: crate::open::Options,
     ) -> Result<Self, Error> {
-        create_opts.destination_must_be_empty = true;
         let mut repo = crate::ThreadSafeRepository::init_opts(path, kind, create_opts, open_opts)?.to_thread_local();
         url.canonicalize(repo.options.current_dir_or_empty())
```
[the remark this answers is not in the document]
Comment (j-walther) [reply gix/src/clone/mod.rs:106]: Done

File: gix/src/lib.rs [hunk 2]
```diff
@@ -293,7 +293,15 @@ pub fn open_with_environment_overrides(directory: impl Into<std::path::PathBuf>)
...
 pub fn init(directory: impl AsRef<std::path::Path>) -> Result<Repository, init::Error> {
-    ThreadSafeRepository::init(directory, create::Kind::WithWorktree, create::Options::default()).map(Into::into)
+    ThreadSafeRepository::init(
+        directory,
+        create::Kind::WithWorktree,
+        create::Options {
+            destination_must_be_empty: false,
+            ..Default::default()
+        },
+    )
```
Comment (j-walther) [inline gix/src/lib.rs:303]: Restores previous behaviour which was also enforced by the `init_into_non_empty_directory_is_allowed_by_default` test.

**Comment by j-walther** [comment]
Hi @Byron, this is rebased on top of `main` and addresses the open Drop-safety concern that was raised in both your review and the Copilot reviewer's follow-ups.

Changes since you last looked:

- **Drop safety fix** ([`36f3805`](https://github.com/GitoxideLabs/gitoxide/pull/2508/commits/36f3805f0)): `PrepareFetch` and `PrepareCheckout` now track whether the destination directory was empty at construction time. On drop, only the `.git` directory we created is removed when the destination held pre-existing user files — `remove_dir_all(workdir)` is reserved for the empty-destination case. This directly addresses your concern that opting in to non-empty destinations could otherwise wipe data on a failed clone.
- **Regression test** for that exact failure path (`drop_after_failed_fetch_into_non_empty_directory_preserves_pre_existing_files` in `gix/tests/gix/clone.rs`) — I verified it fails without the Drop fix and passes with it.
- **Field doc updated** on `Options::destination_must_be_empty` to spell out the `None` / `Some(true)` / `Some(false)` semantics.
- Rebased onto `main` (clean except for two small test-file conflicts already resolved).

The `bool` → `Option<bool>` change is intentional and the cleanest way I found to let `gix::init` keep its documented permissive default while `prepare_clone` stays strict by default. Happy to revisit the API shape if you'd prefer a different split.

CI is green.

File: gix/src/clone/mod.rs [hunk 3]
```diff
@@ -99,10 +99,9 @@ impl PrepareFetch {
         mut url: gix_url::Url,
         path: &std::path::Path,
         kind: crate::create::Kind,
-        mut create_opts: crate::create::Options,
+        create_opts: crate::create::Options,
         open_opts: crate::open::Options,
     ) -> Result<Self, Error> {
```
[the remark this answers is not in the document]
Comment (j-walther) [reply gix/src/clone/mod.rs]: Fixed in 36f3805. `PrepareFetch` now records whether the destination was empty at construction and only removes the `.git` directory on drop when it wasn't. New regression test `drop_after_failed_fetch_into_non_empty_directory_preserves_pre_existing_files` covers the failure-rollback path explicitly.

File: gix/src/clone/mod.rs [hunk 4]
```diff
@@ -102,7 +102,10 @@ impl PrepareFetch {
         mut create_opts: crate::create::Options,
         open_opts: crate::open::Options,
     ) -> Result<Self, Error> {
-        create_opts.destination_must_be_empty = true;
+        if create_opts.destination_must_be_empty.is_none() {
+            create_opts.destination_must_be_empty = Some(true);
+        }
```
[the remark this answers is not in the document]
Comment (j-walther) [reply gix/src/clone/mod.rs:109]: Addressed in 36f3805 — see the new `destination_was_empty` field on `PrepareFetch`/`PrepareCheckout` and the updated `Drop` impls in `gix/src/clone/access.rs` and `gix/src/clone/checkout.rs`. The dropped handle only `remove_dir_all`s the workdir when it was empty at construction; otherwise it only removes the `.git` directory we created.

File: gix/tests/gix/clone.rs [hunk 5]
```diff
@@ -550,6 +550,36 @@ mod blocking_io {
...
+            gix::create::Kind::WithWorktree,
+            gix::create::Options {
+                destination_must_be_empty: Some(false),
+                ..Default::default()
+            },
+            restricted(),
+        )?;
+        let (mut checkout, _out) =
+            prepare.fetch_then_checkout(gix::progress::Discard, &std::sync::atomic::AtomicBool::default())?;
+        let (repo, _) = checkout.main_worktree(gix::progress::Discard, &std::sync::atomic::AtomicBool::default())?;
```
[the remark this answers is not in the document]
Comment (j-walther) [reply gix/tests/gix/clone.rs:609]: Added in 36f3805 as `drop_after_failed_fetch_into_non_empty_directory_preserves_pre_existing_files`. It seeds a pre-existing file, triggers a deterministic fetch failure via a non-existing ref, drops the prepare handle, and asserts both that the pre-existing file is intact and that the `.git` directory is cleaned up.

File: gix/src/create.rs [hunk 6]
```diff
@@ -113,7 +113,7 @@ pub struct Options {
     /// If true, and the kind of repository to create has a worktree, then the destination directory must be empty.
     ///
     /// By default repos with worktree can be initialized into a non-empty repository as long as there is no `.git` directory.
```
[the remark this answers is not in the document]
Comment (j-walther) [reply gix/src/create.rs]: Doc updated in 36f3805 to describe the `None` / `Some(true)` / `Some(false)` cases explicitly, including the note that `PrepareFetch::new` interprets `None` as strict to preserve historical clone behavior.

File: gix/src/create.rs [hunk 7]
```diff
@@ -119,6 +119,15 @@ pub struct Options {
     pub fs_capabilities: Option<gix_fs::Capabilities>,
 }
 
+impl Default for Options {
+    fn default() -> Self {
+        Options {
+            destination_must_be_empty: true,
+            fs_capabilities: None,
+        }
```
[the remark this answers is not in the document]
Comment (j-walther) [reply gix/src/create.rs]: Aligned in 36f3805 — the field default reverts to `None`/permissive (matches the doc), and `PrepareFetch::new` is the only place that flips `None` → `Some(true)` so clones stay strict. The Default impl is back to derived.

File: gix/src/create.rs [hunk 8]
```diff
@@ -113,7 +113,7 @@ pub struct Options {
     /// If true, and the kind of repository to create has a worktree, then the destination directory must be empty.
     ///
     /// By default repos with worktree can be initialized into a non-empty repository as long as there is no `.git` directory.
-    pub destination_must_be_empty: bool,
+    pub destination_must_be_empty: Option<bool>,
```
[the remark this answers is not in the document]
Comment (j-walther) [reply gix/src/create.rs:127]: Acknowledged — this is an intentional public API change so `gix::init` and `prepare_clone` can have different effective defaults without forcing every caller to opt out.

File: gix/src/create.rs [hunk 9]
```diff
@@ -135,7 +135,7 @@ pub fn into(
     let mut dot_git = directory.into();
     let bare = matches!(kind, Kind::Bare);
 
-    if bare || destination_must_be_empty {
+    if bare || destination_must_be_empty.unwrap_or_default() {
```
[the remark this answers is not in the document]
Comment (j-walther) [reply gix/src/create.rs]: Already addressed in [a71f6ad9](https://github.com/GitoxideLabs/gitoxide/pull/2508/commits/a71f6ad9e) — switched to `destination_must_be_empty.unwrap_or(false)` per your suggestion.

File: gix/src/clone/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (j-walther) [reply gix/src/clone/mod.rs:121]: The PR title was updated to *"Allow checkouts of repositories into non-empty directories"* shortly after that comment.

**Review by Byron — APPROVED** [review]
This PR got lucky as I nerd-sniped myself into probably putting more time in it as it would have cost me if I would have implemented this feature from the start.

Nowadays, I think it costs more to lift a mediocre PR into what's needed than to do it from scratch given agent support, and it's something I will consider in future, which probably leans more towards closing such PRs than engaging with it.

Lesson learned, I hope.

File: gix/src/clone/mod.rs [hunk 10]
```diff
@@ -102,7 +104,17 @@ impl PrepareFetch {
...
+
+        // Capture this before init_opts creates `.git`, otherwise the check below would see our own files.
+        let remove_worktree_on_drop = match std::fs::read_dir(path) {
+            Ok(mut entries) => entries.next().is_none(),
+            // Non-existent (or unreadable) — init_opts will create the directory.
+            Err(_) => true,
```
[the remark this answers is not in the document]
Comment (Byron) [reply gix/src/clone/mod.rs]: Great catch! Strange I didn't pick up on this.

File: gix/src/create.rs [hunk 11]
```diff
@@ -108,12 +108,21 @@ fn create_dir(p: &Path) -> Result<(), Error> {
...
-    /// By default repos with worktree can be initialized into a non-empty repository as long as there is no `.git` directory.
-    pub destination_must_be_empty: bool,
+    /// - `None` (default): initialize like Git and allow a non-empty destination directory, as long as no `.git`
+    ///   directory is present.
+    /// - `Some(true)`: require an empty destination directory.
+    /// - `Some(false)`: explicitly allow initialization into a non-empty destination directory (still requires that no
+    ///   `.git` directory is present).
+    ///
+    /// When this is `Some(false)|None` for clones, checkout failure cleanup preserves the entire destination, including the
+    /// created `.git` directory, so users can inspect or clean up the partially cloned repository themselves.
```
[the remark this answers is not in the document]
Comment (Byron) [reply gix/src/create.rs]: Great catch!

File: gix/src/init.rs [hunk 12]
```diff
@@ -42,8 +42,9 @@ pub enum Error {
 impl ThreadSafeRepository {
     /// Create a repository with work-tree within `directory`, creating intermediate directories as needed.
     ///
-    /// Fails without action if there is already a `.git` repository inside of `directory`, but
-    /// won't mind if the `directory` otherwise is non-empty.
+    /// Fails without action if the destination directory isn't empty unless
+    /// [`create::Options::destination_must_be_empty`][crate::create::Options::destination_must_be_empty] is `None`
+    /// or `Some(false)`.
```
[the remark this answers is not in the document]
Comment (Byron) [reply gix/src/init.rs]: Right!

File: gix/tests/gix/clone.rs [hunk 13]
```diff
@@ -585,6 +588,103 @@ mod blocking_io {
...
+
+        assert_eq!(
+            std::fs::read(&existing_path)?,
+            EXISTING_CONTENT,
+            "pre-existing user files must survive a failed clone+drop"
+        );
+        assert!(
+            destination.join(".git").is_dir(),
+            "the .git directory we created should remain for user cleanup"
+        );
```
[the remark this answers is not in the document]
Comment (Byron) [reply gix/tests/gix/clone.rs:725]: Ignore the description
